### PR TITLE
If user tries to revoke group's access to all tables, revoke access to raw queries as well

### PR DESCRIPTION
**Pull request by attekei**
**TLDR: When you change permissions on table or schema level, now the permissions UI tries to infer the permissions for the ancestors (both schema&db permissions if you are editing table permissions, only db permissions if you are editing schema permissions) accordingly based on their descendant tables. This should prevent the permissions tree from getting to an invalid state.**

Addresses #4273 by revoking an access to raw queries at the same time when user revokes an access to the very last table in the database that still has unrestricted access. See this screenshot for the new new warning dialog:

screen shot 2017-04-24 at 9 17 25

**TODO:**
- [x] Trigger the warning dialog when user tries to revoke access to last table
- [x] Revoke the access to raw queries properly after user has confirmed the revoking

### CODE REVIEW

**Comment by attekei** [comment]
I figured out what causes #4970. These issues are related so I will tackle both in this PR.

**Comment by attekei** [comment]
Added an experimental way for inferring db/schema permissions based on their descendant tables which addresses #4970. I don't know is this an optimal approach so I should talk about this with @camsaul and others on Monday.

**Comment by attekei** [comment]
Ready for review!

**Review by camsaul — APPROVED** [review]
sounds like it makes sense

**Review by tlrobinson — APPROVED** [review]
Seems reasonable.

It's unfortunate we have to duplicate so much logic for this on the frontend and backend, but it's sort of required for the "batch" permission update model we chose.

**Review by mazameli — APPROVED** [review]
I couldn't break anything, so seems good to me.

**Comment by attekei** [comment]
In terms of functionality this is ready to be merged, but now I think that I should still write some unit test for this. The permission functionality is pretty complex/fragile and it has zero test coverage so it's time to fix that.

**Comment by attekei** [comment]
I will merge this to `release-0.24.0` and create a separate PR for tests.
